### PR TITLE
Notes on using git

### PR DESCRIPTION
**Pull request by richvdh**
I wrote down some stuff about git.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -98,24 +96,55 @@ in the format of `PRnumber.type`. The type can be one of the following:
...
 contributions and would like to have you shouted out in the release notes!
 
 For example, a fix in PR #1234 would have its changelog entry in
-`changelog.d/1234.bugfix`, and contain content like "The security levels of
-Florbs are now validated when received over federation. Contributed by Jane
-Matrix.".
+`changelog.d/1234.bugfix`, and contain content like:
+
+> The security levels of Florbs are now validated when received over
+> via the `/federation/florb` endpoint. Contributed by Jane Matrix.
```
Comment (clokep) [inline CONTRIBUTING.md:113]:
```suggestion
> The security levels of Florbs are now validated when received
> via the `/federation/florb` endpoint. Contributed by Jane Matrix.
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -98,24 +96,55 @@ in the format of `PRnumber.type`. The type can be one of the following:
...
+chicken-and-egg problem.
+
+There are two options for solving this:
+
+ 1. Open the PR without a changelog file, see what number you got, and *then*
+    add the changelog file to your branch (see [Updating your pull
+    request](#updating-your-pull-request)), or:
 
-## Debian changelog
+ 1. look at the [list of all
```
Comment (clokep) [inline CONTRIBUTING.md:132]:
```suggestion
 1. Look at the [list of all
```

File: docs/dev/git.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+One counterargument is that it is sometimes useful to see how a PR evolved as
+it went through review cycles. This is true, but that information is always
+available via the GitHub UI (or via the little-known [refs/pull
+namespace](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/checking-out-pull-requests-locally)).
+
+
+Of course, in reality, things are more complicated than that: we have release
+branches as well as `develop` and `master` and we deliberately merge changes
+between them; often slip through and have to be fixed later. So this is very
```
Comment (clokep) [inline docs/dev/git.md:52]: I'm having trouble following the first sentence here. I get the gist but it is hard to read.

File: docs/dev/git.md [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+much an ideal that we're aiming for rather than a cast-iron set of rules which
+must be obeyed.
+
+Merges, squashes, rebases: wtf?
+-------------------------------
+
+Ok, so that's what we'd like to achieve. How do we achieve it?
+
+The TL;DR is: when you come to merge a pull request, you *probably* want to
+[“squash and merge”](git/squash.png).
```
Comment (clokep) [inline docs/dev/git.md:62]: Did you want to show this picture on this page?

File: docs/dev/git.md [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+On keeping the commit history clean
+-----------------------------------
+
+In an ideal world, our git commit history would be a linear progression of
+commits each of which contains a single change building on what came
+before. Here, by way of an arbitrary example, is the top of `git log --graph
+b2dba0607`:
+
+<img src="git/clean.png" alt="clean git graph" width="500px">
```
Comment (clokep) [inline docs/dev/git.md:12]: I'm assuming this uses HTML instead of Markdown here so you can control the width?

File: docs/dev/git.md [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+or a set of changes to (some subset of) the world: for example, this happens
+when a PR is landed, or as part of our release process.
+
+So, what counts as a more- or less-stable branch? A little reflection will show
+that our active branches are ordered thus, from more-stable to less-stable:
+
+ * `master` (tracks our last release).
+ * `release-vX.Y.Z` (the branch where we prepare the next release)<sup
+   id="a3">[3](#f3)</sup>.
+ * PR branches which are targetting the release.
```
Comment (clokep) [inline docs/dev/git.md:125]: Is targetting (as opposed to targeting) proper British spelling? The Internet seems to be unclear to me here.

File: docs/dev/git.md (hunk 5 above)
Comment (richvdh) [reply docs/dev/git.md:12]: yup.

File: docs/dev/git.md (hunk 6 above)
Comment (richvdh) [reply docs/dev/git.md:125]: oxford dictionary says not. I think I was extrapolating from "forgetting"...

File: docs/dev/git.md (hunk 4 above)
Comment (richvdh) [reply docs/dev/git.md:62]: yes, probably.

File: docs/dev/git.md (hunk 3 above)
Comment (richvdh) [reply docs/dev/git.md]: Indeed. I added the missing word and reworded a bit. how is it now?

Comment (clokep) [reply docs/dev/git.md]: Looks good now! Thanks!

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -1,19 +1,20 @@
-# Contributing code to Matrix
+# Contributing code to Synapse
```
[the remark this answers is not in the document]
Comment (richvdh) [reply CONTRIBUTING.md:1]: ok, please could they not? increasingly what is here is specific to synapse.

**Pull request merged**